docs/markdown/Ad-Hoc Tools/adhoc-tool.md

```diff
@@ -0,0 +1,86 @@
+
+`adhoc_tool` provides you with the building blocks needed to put together a custom build process without needing to develop and maintain a plugin. The level of initial effort involved in using `adhoc_tool` is significantly lower than that of [writing a plugin](doc:plugins-overview), so it's well-suited to consuming one-off scripts, or for rapidly prototyping a process before actually writing a plugin. The tradeoff is that there is more manual work involved in defining your project structure, and that the targets that define the tools you consume are less easy to reuse.
+
+The `antlr` demo in the [`example-adhoc` respository](https://github.org/pantsbuild/example-adhoc) shows an example of running a JVM-based tool to transparently generate Python code that can be used in another language:
+
+```
+adhoc_tool(
+    name="run_antlr",
+    runnable=":antlr4",
+    args=["Expr.g4", "-Dlanguage=Python3", "-o", "expr_parser", "-package", "expr_parser",],
+    output_directories=["expr_parser",],
+    # These are consumed by `antlr`, but are not relevant to this target's dependents.
+    execution_dependencies=[":grammars"],
+    # These are needed by the code that is output by this target
+    output_dependencies=[":antlr4-python3-runtime",],
+    root_output_directory=".",
+    log_output=True,
+)
+```
+
+## `runnable` targets
+
+"Runnable" targets are targets that Pants knows how to execute within its sandbox. Generally, these correspond to targets that can be executed with the `pants run` goal, and include first-party source files, as well as third-party dependencies.
+
+The tool will be run with values from `args` specified as arguments. By default, the process' working directory will be the directory where the `BUILD` file is defined. This can be adjusted using the `workdir` field.
+
+
+### Specifying dependencies
+
+`adhoc_tool` has more complexity surrounding dependencies compared with Pants' first-class targets. This is because you need to do manual work to set up the execution environment, which is usually taken care of by plugin code.
+
+`adhoc_tool` has three dependencies fields:
+
+* `output_dependencies`, which defines dependencies that are required to effectively consume the output of the tool, _e.g._ runtime libraries for generated code bindings. These are returned when resolving the transitive dependencies of any (transitive) dependents of the `adhoc_tool` target.
+* `execution_dependencies`, which define data dependencies required for the tool to produce its output. These are not considered when resolving transitive dependencies that include this `adhoc_tool` target.
+* `runnable_dependencies`, which define runnables that the `adhoc_tool` needs to execute as a subprocess. These are also not considered when resolving transitive dependencies. The discussion of `system_binary` later in this page shows one key use of `runnable_dependencies`.
```

Comment:
Even if we don't go with my table idea above, I think it'd be good to discuss how a runnable is introduced because AIUI, it's a bit different to 'usual'. 

(My understanding is that it's plopped into an unpredictable location but placed on `PATH` under its target name, meaning, for instance, a PEX should be run via `target-name` directly, rather than its usual `{chroot}/path.to.target/target-name.pex` position, right?)

Reply:
I suspect that if you're needing a `pex` as a `runnable_dependency`, your processes aren't as decomposed as they probably should be.

Reply:
Hehe, that's fair.